src/socket.c

```diff
@@ -52,68 +55,57 @@ static int bind_socket(const fastd_bind_address_t *addr) {
 	fastd_setnonblock(fd);
 #endif
 
-	int one = 1;
-
 #ifdef USE_PKTINFO
-	if (setsockopt(fd, IPPROTO_IP, IP_PKTINFO, &one, sizeof(one))) {
+	if (af == AF_INET && setsockopt(fd, IPPROTO_IP, IP_PKTINFO, &one, sizeof(one))) {
 		pr_error_errno("setsockopt: unable to set IP_PKTINFO");
 		goto error;
 	}
 #endif
 
+	if (af == AF_INET6 && setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one))) {
+		pr_error_errno("setsockopt: unable to set IPV6_RECVPKTINFO");
+		goto error;
+	}
+
 #ifdef USE_FREEBIND
+#ifdef USE_FREEBIND6
+	if (af == AF_INET && setsockopt(fd, IPPROTO_IP, IP_FREEBIND, &one, sizeof(one)))
+		pr_warn_errno("setsockopt: unable to set IP_FREEBIND");
+	if (af == AF_INET6 && setsockopt(fd, IPPROTO_IPV6, IPV6_FREEBIND, &one, sizeof(one)))
+		pr_warn_errno("setsockopt: unable to set IPV6_FREEBIND");
+#else
 	if (setsockopt(fd, IPPROTO_IP, IP_FREEBIND, &one, sizeof(one)))
 		pr_warn_errno("setsockopt: unable to set IP_FREEBIND");
 #endif
-
-	if (af == AF_INET6) {
-		if (setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one))) {
-			pr_error_errno("setsockopt: unable to set IPV6_RECVPKTINFO");
-			goto error;
-		}
-	}
+#endif
 
 #ifdef USE_BINDTODEVICE
-	if (addr->bindtodev && !fastd_peer_address_is_v6_ll(&addr->addr)) {
-		if (setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, addr->bindtodev, strlen(addr->bindtodev))) {
-			pr_warn_errno("setsockopt: unable to bind to device");
-			goto error;
-		}
+	if (addr->bindtodev && !fastd_peer_address_is_v6_ll(&addr->addr) &&
+			setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, addr->bindtodev, strlen(addr->bindtodev))) {
```

Comment:
In my opinion changing the nested `if` to a single condition that mixes condition and side-effects makes the code unclearer.